Fixing Prefabs so FusedBombs can spawn lit

joshuaskelly opened the pull request:
# Summary
Fixes bombs never spawning lit. This was because `Level.addEntity` always set source to `Source.LEVEL_START` and the FusedBomb class has a guard against that case.

I added a `Level.addEntity(Entity e, Source source)` so prefabs can set the source correctly.

Interrupt commented on Dungeoneer/src/com/interrupt/dungeoneer/entities/items/Wand.java:
```diff
@@ -194,7 +194,7 @@ public void tickEquipped(Player player, Level level, float delta, String equipLo
 				chargeEffect.tex = 0;
 
 				chargeTime = 0f;
-				level.addEntity(chargeEffect);
```
FYI, the Level class has a SpawnEntity method that does this same thing

Interrupt commented on Dungeoneer/src/com/interrupt/dungeoneer/entities/Prefab.java:
```diff
@@ -49,7 +49,7 @@ public void SpawnEntity(Level level, Entity e) {
 		if (e instanceof Monster) {
 			((Monster)e).Init(level, Game.instance.player.level);
 		}
-		level.addEntity(e);
```
Changing the Prefab behavior is potentially scary, could this be wired up by overriding the `tossItem` function like the Potion does?

https://github.com/Interrupt/delverengine/blob/18a41f22f4e72e2970e84ea14e6cae6722b42de4/Dungeoneer/src/com/interrupt/dungeoneer/entities/items/Potion.java#L237

joshuaskelly replied:
This is for when you break a pot and bomb drops. Instead of changing how prefabs spawn things, I could change the guard inside of `FusedBomb`

joshuaskelly commented:
Reverted the changes to `Level` and `Prefab`. I just removed the guard in `FusedBomb.init()`

Previous:
```java
if(source == Level.Source.SPAWNED && !wasSpawned) {
    isLit = Game.rand.nextFloat() < this.chanceIsLit;
    isDud = Game.rand.nextFloat() < this.chanceIsDud;
}
else if(source == Level.Source.LEVEL_START) {	
    isLit = false;	
    isDud = Game.rand.nextFloat() < this.chanceIsDud;	
}	

wasSpawned = true;
```

New:
```java
isLit = Game.rand.nextFloat() < chanceIsLit;
isDud = Game.rand.nextFloat() < chanceIsDud;
wasSpawned = true;
```

PythooonUser reviewed (approved):
Works as expected.

Interrupt commented on Dungeoneer/src/com/interrupt/dungeoneer/entities/items/FusedBomb.java:
```diff
@@ -9,186 +9,176 @@
-			isLit = false;
-			isDud = Game.rand.nextFloat() < this.chanceIsDud;
-		}
+        isLit = Game.rand.nextFloat() < chanceIsLit;
```
Won't this now cause the bomb logic to reset on save loads, now that it isn't guarding on `wasSpawned`?

joshuaskelly replied:
Yes, yes you are correct. I've reworked the `init` method:
```java
// Don't set any fields if this was placed in the editor.
if (wasJustPlacedInEditor(source)) {
    return;
}

// Guard against relighting on loading of a save.
if (wasSpawned) {
    return;
}

isLit |= Game.rand.nextFloat() < chanceIsLit;
isDud |= Game.rand.nextFloat() < chanceIsDud;
wasSpawned = true;
```
